In [1]:
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

# Data cleaning

## Loading the CSV file

In [2]:
path = ""
file_name = "OpenFood.csv"
data_path = path+file_name
df = pd.read_csv(data_path, sep = '\t')

/Users/Mathias/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (0,3,5,24,25,26,28,36,41,49) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


## DataFrame Exploration

In [3]:
df.head()

code                                                url   creator  \
0  0000000000017  http://world-en.openfoodfacts.org/product/0000...   kiliweb   
1  0000000000031  http://world-en.openfoodfacts.org/product/0000...  isagoofy   
2  0000000000123  http://world-en.openfoodfacts.org/product/0000...   kiliweb   
3  0000000000291  http://world-en.openfoodfacts.org/product/0000...   kiliweb   
4  0000000000949  http://world-en.openfoodfacts.org/product/0000...   kiliweb   

    created_t      created_datetime last_modified_t last_modified_datetime  \
0  1529059080  2018-06-15T10:38:00Z      1529059204   2018-06-15T10:40:04Z   
1  1539464774  2018-10-13T21:06:14Z      1539464817   2018-10-13T21:06:57Z   
2  1535737982  2018-08-31T17:53:02Z      1535737986   2018-08-31T17:53:06Z   
3  1534239669  2018-08-14T09:41:09Z      1534239732   2018-08-14T09:42:12Z   
4  1523440813  2018-04-11T10:00:13Z      1523440823   2018-04-11T10:00:23Z   

                product_name generic_name quantity      ...        \
0           Vitória crackers          NaN      NaN      ...         
1                      Cacao          NaN    130 g      ...         
2      Sauce Sweety chili 0%          NaN      NaN      ...         
3                  Mendiants          NaN      NaN      ...         
4  Salade de carottes râpées          NaN      NaN      ...         

  carbon-footprint_100g nutrition-score-fr_100g nutrition-score-uk_100g  \
0                   NaN                     NaN                     NaN   
1                   NaN                     NaN                     NaN   
2                   NaN                     NaN                     NaN   
3                   NaN                     NaN                     NaN   
4                   NaN                     NaN                     NaN   

  glycemic-index_100g water-hardness_100g choline_100g phylloquinone_100g  \
0                 NaN                 NaN          NaN                NaN   
1                 NaN                 NaN          NaN                NaN   
2                 NaN                 NaN          NaN                NaN   
3                 NaN                 NaN          NaN                NaN   
4                 NaN                 NaN          NaN                NaN   

  beta-glucan_100g inositol_100g carnitine_100g  
0              NaN           NaN            NaN  
1              NaN           NaN            NaN  
2              NaN           NaN            NaN  
3              NaN           NaN            NaN  
4              NaN           NaN            NaN  

[5 rows x 173 columns]

### DataFrame properties

Are there some duplicated rows?

In [4]:
np.sum(df.duplicated().values == True)

27

So the duplicated values are deleted:

In [5]:
df.drop_duplicates(inplace = True)

We can also drop the column that only conain NaN values:

In [6]:
df.dropna(axis = 1, how='all', inplace = True)

In [7]:
df.shape

(683257, 169)

So now there are 683257 rows and 196 columns

Now can the code columnm be set as the index. In order to be so it must be unique:

In [29]:
df.code.is_unique

False

The codes are not unique so let's check what are those values are:

In [27]:
duplicate_codes = np.where(df.duplicated('code', keep=False).values)[0]
dups = df.iloc[duplicate_codes]
dups.code = dups.code.astype(float)
dups.sort_values(by = 'code')

/Users/Mathias/anaconda3/lib/python3.6/site-packages/pandas/core/generic.py:4401: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


code                                                url  \
207855  1.160220e+05  http://world-en.openfoodfacts.org/product/1160...   
10832   1.160220e+05  http://world-en.openfoodfacts.org/product/0011...   
74720   6.356460e+05  http://world-en.openfoodfacts.org/product/0063...   
147293  6.356460e+05  http://world-en.openfoodfacts.org/product/0635...   
74722   6.356840e+05  http://world-en.openfoodfacts.org/product/0063...   
147299  6.356840e+05  http://world-en.openfoodfacts.org/product/0635...   
89032   7.228100e+05  http://world-en.openfoodfacts.org/product/0072...   
163560  7.228100e+05  http://world-en.openfoodfacts.org/product/0722...   
186524  8.379960e+05  http://world-en.openfoodfacts.org/product/0837...   
111770  8.379960e+05  http://world-en.openfoodfacts.org/product/0083...   
203515  8.887070e+05  http://world-en.openfoodfacts.org/product/0888...   
116859  8.887070e+05  http://world-en.openfoodfacts.org/product/0088...   
117418  8.984780e+05  http://world-en.openfoodfacts.org/product/0089...   
206086  8.984780e+05  http://world-en.openfoodfacts.org/product/0898...   
140237  3.318233e+06  http://world-en.openfoodfacts.org/product/0331...   
357256  3.318233e+06  http://world-en.openfoodfacts.org/product/3318...   
510594  4.159932e+06  http://world-en.openfoodfacts.org/product/4159...   
141140  4.159932e+06  http://world-en.openfoodfacts.org/product/0415...   
144862  6.136567e+06  http://world-en.openfoodfacts.org/product/0613...   
576411  6.136567e+06  http://world-en.openfoodfacts.org/product/6136...   
207518  1.111082e+10  http://world-en.openfoodfacts.org/product/1111...   
4272    1.111082e+10  http://world-en.openfoodfacts.org/product/0011...   
12091   1.356200e+10  http://world-en.openfoodfacts.org/product/0013...   
208756  1.356200e+10  http://world-en.openfoodfacts.org/product/1356...   
208757  1.356200e+10  http://world-en.openfoodfacts.org/product/1356...   
12092   1.356200e+10  http://world-en.openfoodfacts.org/product/0013...   
208942  1.410007e+10  http://world-en.openfoodfacts.org/product/1410...   
12480   1.410007e+10  http://world-en.openfoodfacts.org/product/0014...   
208943  1.410010e+10  http://world-en.openfoodfacts.org/product/1410...   
12527   1.410010e+10  http://world-en.openfoodfacts.org/product/0014...   
...              ...                                                ...   
262035           NaN                                                NaN   
262046           NaN                                                NaN   
262070           NaN                                                NaN   
262083           NaN                                                NaN   
262091           NaN                                                NaN   
262093           NaN                                                NaN   
262099           NaN                                                NaN   
262146           NaN                                                NaN   
262166           NaN                                                NaN   
262176           NaN                                                NaN   
262178           NaN                                                NaN   
262182           NaN                                                NaN   
262184           NaN                                                NaN   
262195           NaN                                                NaN   
262197           NaN                                                NaN   
262204           NaN                                                NaN   
262207           NaN                                                NaN   
262243           NaN                                                NaN   
262284           NaN                                                NaN   
262287           NaN                                                NaN   
262317           NaN                                                NaN   
262319           NaN                     

As it can be seen there are two types of duplicates:
Some codes appear twice and the some code are NaN. As it can be observed when code are duplicated they are often the same and sometimes one row contain more information than the other. So we could just keep the rows that contain the most information. And for the rows with NaN the values of the columns are not matching the column nae so there must be a shift of values. As a lot of values are missing, it would be complex to recover the true data and there are only tens of code with NaNs, we can delete them:

In [32]:
df = df[df.code.astype(float) != float('nan')]
